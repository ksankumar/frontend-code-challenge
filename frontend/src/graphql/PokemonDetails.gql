query PokemonDetails($id: ID!) {
  pokemonById(id: $id) {
    id
    name
    weight {
      maximum
      minimum
    }
    height {
      maximum
      minimum
    }
    types
    resistant
    attacks {
      fast {
        name
        type
        damage
      }
      special {
        name
        type
        damage
      }
    }
    weaknesses
    fleeRate
    maxCP
    evolutions {
      id
      name
    }
    evolutionRequirements {
      amount
      name
    }
    maxHP
    image
    sound
    isFavorite
  }
}
