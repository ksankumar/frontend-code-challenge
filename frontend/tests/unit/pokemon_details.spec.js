import { config, shallowMount, createLocalVue } from "@vue/test-utils";
import { createMockClient } from "mock-apollo-client";
import Vue from "vue";
import Vuetify from "vuetify";
import VueRouter from "vue-router";
import VueApollo from "vue-apollo";
import mockPokemonDetails from "../data/details.json";
import mockPokemonList from "../data/list.json";
import pokemonDetailsQuery from "@/graphql/pokemonDetails.query.gql";
import favoritePokemonMutation from "@/graphql/favoritePokemon.mutation.gql";
import unFavoritePokemonMutation from "@/graphql/unFavoritePokemon.mutation.gql";
import localTestQuery from "@/graphql/localTest.query.gql";
import PokemonDetails from "@/views/Details";

config.showDeprecationWarnings = false;
const localVue = createLocalVue();

const routes = [
  {
    path: "/details/123",
    name: "PokemonDetails",
    component: PokemonDetails
  }
];
const router = new VueRouter({ routes });
Vue.use(Vuetify);
Vue.use(VueRouter);
Vue.use(VueApollo);

describe("Pokemon component", () => {
  let wrapper;
  let mockClient;
  let apolloProvider;
  let requestHandlers;
  const createComponent = (handlers, data) => {
    // We want to specify default handlers for different operations and overwrite them with passed handlers
    requestHandlers = {
      pokemonDetailsQueryHandler: jest
        .fn()
        .mockResolvedValue(mockPokemonDetails),
      favoritePokemonMutationHandler: jest
        .fn()
        .mockResolvedValue(mockPokemonList),
      unFavoritePokemonMutationHandler: jest
        .fn()
        .mockResolvedValue(mockPokemonList),
      ...handlers
    };

    mockClient = createMockClient({
      resolvers: {}
    });

    mockClient.cache.writeQuery({
      query: localTestQuery,
      data: {
        fetchLocalUser: {
          __typename: "Pokemon",
          name: "Test"
        }
      }
    });

    // Unfortunately, it's not possible to override handler already set
    // So we need to do all the work for setting handlers in the factory
    mockClient.setRequestHandler(
      pokemonDetailsQuery,
      requestHandlers.pokemonDetailsQueryHandler
    );
    mockClient.setRequestHandler(
      favoritePokemonMutation,
      requestHandlers.favoritePokemonMutationHandler
    );
    mockClient.setRequestHandler(
      unFavoritePokemonMutation,
      requestHandlers.unFavoritePokemonMutationHandler
    );

    apolloProvider = new VueApollo({
      defaultClient: mockClient
    });

    wrapper = shallowMount(PokemonDetails, {
      localVue,
      apolloProvider,
      router,
      data() {
        return {
          ...data
        };
      }
    });
    wrapper.vm.$route.params["id"] = "001";
  };

  it("renders a pokemon details component", async () => {
    createComponent();
    expect(wrapper.is(PokemonDetails)).toBe(true);
    expect(wrapper.vm).toBeTruthy();
    expect(
      wrapper.findComponent({ name: PokemonDetails.name }).exists()
    ).toBeTruthy();
    expect(requestHandlers.pokemonDetailsQueryHandler).toHaveBeenCalled();
  });

  it("renders message when component is created", () => {
    expect(wrapper.text()).toContain("Bulbasaur");
  });

  it("displayed pokemon correctly with query data", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("make a un-favorite pokemon", async () => {
    wrapper.vm.makeFavorite(mockPokemonDetails.data.pokemonById);
    expect(
      requestHandlers.unFavoritePokemonMutationHandler
    ).toHaveBeenCalledWith({
      id: "001"
    });
    expect(wrapper.html()).toContain("mdi-heart");
  });

  it("make a favorite pokemon", async () => {
    wrapper.vm.makeFavorite({ id: "001", isFavorite: false });
    expect(requestHandlers.favoritePokemonMutationHandler).toHaveBeenCalledWith(
      {
        id: "001"
      }
    );
  });
});
