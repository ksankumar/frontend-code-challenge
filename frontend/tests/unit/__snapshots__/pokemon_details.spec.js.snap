// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon component displayed pokemon correctly with query data 1`] = `
<v-container-stub tag="div">
  <div>
    <v-card-stub loaderheight="4" tag="div" outlined="true"><template></template>
      <v-img-stub height="400" contain="true" options="[object Object]" position="center center" src="https://img.pokemondb.net/artwork/bulbasaur.jpg" transition="fade-transition">
        <audio-player-stub audiosrc="http://localhost:4000/sounds/1"></audio-player-stub>
      </v-img-stub>
      <v-sheet-stub color="secondary" tag="div">
        <v-card-stub loaderheight="4" tag="div" color="transparent" tile="true" flat="true">
          <v-list-item-stub activeclass="" tag="div" threeline="true">
            <v-list-item-content-stub>
              <v-list-item-title-stub class="headline font-weight-bold">Bulbasaur
              </v-list-item-title-stub>
              <v-list-item-subtitle-stub class="font-weight-bold">Grass, Poison
              </v-list-item-subtitle-stub>
            </v-list-item-content-stub>
            <v-list-item-action-stub>
              <v-btn-stub tag="button" activeclass="" icon="true" type="button">
                <v-icon-stub color="red" size="30">mdi-heart
                </v-icon-stub>
              </v-btn-stub>
            </v-list-item-action-stub>
          </v-list-item-stub>
          <v-card-text-stub class="pt-0">
            <v-row-stub tag="div" nogutters="true" align="center">
              <v-col-stub cols="11" tag="div">
                <v-progress-linear-stub color="#a09ffe" value="100" active="true" buffervalue="100" height="10" rounded="true"></v-progress-linear-stub>
              </v-col-stub>
              <v-col-stub cols="1" tag="div"><b class="font-weight-black float-right">CP:&nbsp;951</b></v-col-stub>
            </v-row-stub>
            <v-row-stub tag="div" nogutters="true" align="center">
              <v-col-stub cols="11" tag="div">
                <v-progress-linear-stub color="#71c1a1" value="100" active="true" buffervalue="100" height="10" rounded="true"></v-progress-linear-stub>
              </v-col-stub>
              <v-col-stub cols="1" tag="div"><b class="font-weight-black float-right">
                  HP:&nbsp;1071
                </b></v-col-stub>
            </v-row-stub>
          </v-card-text-stub>
        </v-card-stub>
        <v-divider-stub></v-divider-stub>
        <v-card-stub loaderheight="4" tag="div" color="transparent" tile="true" flat="true" class="text-center">
          <v-row-stub tag="div" align="center" justify="center">
            <v-col-stub tag="div">
              <v-list-item-content-stub>
                <v-list-item-title-stub class="headline font-weight-bold">
                  Weight
                </v-list-item-title-stub>
                <v-list-item-subtitle-stub class="subtitle-1">
                  6.04kg
                  <span class="px-1">-</span>
                  7.76kg
                </v-list-item-subtitle-stub>
              </v-list-item-content-stub>
            </v-col-stub>
            <v-divider-stub vertical="true"></v-divider-stub>
            <v-col-stub tag="div">
              <v-list-item-content-stub>
                <v-list-item-title-stub class="headline font-weight-bold">
                  Height
                </v-list-item-title-stub>
                <v-list-item-subtitle-stub class="subtitle-1">
                  0.61m
                  <span class="px-1">-</span>
                  0.79m
                </v-list-item-subtitle-stub>
              </v-list-item-content-stub>
            </v-col-stub>
          </v-row-stub>
          <v-divider-stub></v-divider-stub>
        </v-card-stub>
      </v-sheet-stub>
    </v-card-stub>
    <v-fade-transition-stub mode="out-in" origin="top center 0">
      <v-row-stub tag="div">
        <v-col-stub cols="12" sm="4" md="3" tag="div" class="pullUp">
          <apolloquery-stub query="[object Object]" variables="[object Object]" update="function _default(data) {
          return data;
        }" debounce="0" throttle="0" tag="div" prefetch="true" options="[object Object]"></apolloquery-stub>
        </v-col-stub>
        <v-col-stub cols="12" sm="4" md="3" tag="div" class="pullUp">
          <apolloquery-stub query="[object Object]" variables="[object Object]" update="function _default(data) {
          return data;
        }" debounce="0" throttle="0" tag="div" prefetch="true" options="[object Object]"></apolloquery-stub>
        </v-col-stub>
      </v-row-stub>
    </v-fade-transition-stub>
  </div>
</v-container-stub>
`;
